### PR TITLE
Feature 1764 point stat

### LINKED ISSUE #1764: Add ORANK line type to the HiRA output from Point-Stat.
## Describe the Enhancement ##
This request arose via met-help issue from the US Army:
https://rt.rap.ucar.edu/rt/Ticket/Display.html?id=99476

The HiRA algorithm in Point-Stat applies probabilistic and ensemble verification techniques when verifying deterministic model output against point observations. The pseudo-ensemble is defined using forecast values from nearby grid points. So this is the analogue of neighborhood verification with gridded data but for point observations instead. The value at each forecast grid point is replaced by a number between 0 and 1 representing the frequency of a user-configurable event within the neighborhood.

As of met-9.1, the HiRA methodology can produce both probabilistic and ensemble output line types. This task is to also add the ORANK line type as an output for HiRA. During the initial HiRA development this seemed like data overload, but some scientists at the Army would find it useful.

When adding ORANK to the HiRA output from Point-Stat, also consider if any of the other probabilistic and ensemble output line types should be added at the same time.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
No sub-issues.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: Seth
- [x] Select **scientist(s)** or **no scientist** required: no scientist needed

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [x] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- [x] METplus issue to make this output line type configurable: dtcenter/METplus#1103

## Enhancement Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by sethlinden**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

Initial testing was done by running a point_stat command line by hand from kiowa:

./point_stat /d1/personal/linden/git/feature_1764_point_stat/MET/MET_test_data/unit_test/model_data/met_nc/nam/nam_2012040900_F036_APCP24.nc /d1/personal/linden/git/feature_1764_point_stat/MET/test_output/ascii2nc/gauge_2012041012_24hr.nc /d1/personal/linden/git/feature_1764_point_stat/MET/test/config/PointStatConfig_APCP_HIRA -outdir /d1/personal/linden/git/feature_1764_point_stat/MET/test_output/hira -v 3

After adding ORANK to all of the point_stat config files, I ran "make test" to test all calls to point_stat. It was successful. Note, the generation of the ORANK output lines was only turned "on" for PointStatConfig_APCP_HIRA

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Run unit tests for MET (run "make test")

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes or No]**

Yes. I updated all relevant documentation in: docs/Users_Guide/point-stat.rst

- [x] Do these changes include sufficient testing updates? **[Yes or No]**

Yes, I updated all point_stat config files to list ORANK. Tested by running "make test".

- [x] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

Yes, this changes adds ORANK output lines to the output for: PointStatConfig_APCP_HIRA

- [x] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: met/src/tools/core/point_stat/point_stat_conf_info.cc [hunk 1]
```diff
@@ -744,7 +764,7 @@ void PointStatVxOpt::process_config(GrdFileType ftype,
 
    // Conf: output_flag
    output_map = parse_conf_output_flag(&odict, txt_file_type, n_txt);
-
+
```
Comment (georgemccabe) [inline met/src/tools/core/point_stat/point_stat_conf_info.cc]:
```suggestion

```
Remove extra whitespace in blank line

File: met/src/tools/core/point_stat/point_stat.cc [hunk 2]
```diff
@@ -1734,11 +1741,11 @@ void do_hira_ens(int i_vx, const PairDataPoint *pd_ptr) {
...
          continue;
       }
-
+      
       // Write out the ECNT line
       if(conf_info.vx_opt[i_vx].output_flag[i_ecnt] != STATOutputType_None) {
-
-         // Compute ensemble statistics
+	 
+	 //Compute ensemble statistics
```
Comment (georgemccabe) [inline met/src/tools/core/point_stat/point_stat.cc]:
```suggestion
         //Compute ensemble statistics
```
It looks like a mix of tabs and spaces here

**Review by georgemccabe — APPROVED** [review]
The code changes look good to me.

I did notice that there are some blank lines that now have some space characters and some lines that have a mix of tabs and spaces. It would be nice to clean that up, but it won't negatively impact the actual code so it is not imperative. I marked the 2 files with suggested changes to fix these, but I did not mark all of the instances of these cases in those files.

I didn't run the make tests but if you ran them successfully and @JohnHalleyGotway does as well, then I trust that they are correct.

I approve of these changes.

**Review by JohnHalleyGotway — COMMENTED** [review]
I'm going to run a full regression test for this PR. Please don't merge these changes until I get back to you with the results.
Running in kiowa:/d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/test_regression.log
```
mkdir -p /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764
cd /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764
../../../MET/scripts/regression/test_regression.sh feature_1764_point_stat develop > & test_regression.log &
```

**Comment by sethlinden** [comment]
Thanks George for giving this a review. Sorry about the mixture of tabs and spaces, I thought I had my emacs configured properly. I did have to manipulate some of the comments because they would not "tab over" with the backslashes (//). Do you by chance use emacs? Do you have an example .emacs file I can looks at.

**Review by JohnHalleyGotway — CHANGES_REQUESTED** [review]
@sethlinden at least one of the unit tests fails. Looks like you missed adding ORANK to at least one of the Point-Stat config files:
```
CALLING: /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/test/perl/unit.pl  /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/test/xml/unit_met_test_scripts.xml

TEST: test_all_gen_vx_mask_1     - pass -   0.546 sec
TEST: test_all_gen_vx_mask_2     - pass -   0.325 sec
TEST: test_all_pcp_combine_1     - pass -   3.157 sec
TEST: test_all_pcp_combine_2     - pass -   2.855 sec
TEST: test_all_pcp_combine_3     - pass -   5.353 sec
TEST: test_all_pcp_combine_4     - pass -   0.782 sec
TEST: test_all_pcp_combine_5     - pass -   0.717 sec
TEST: test_all_pcp_combine_6     - pass -   1.175 sec
TEST: test_all_mode_1            - pass -   4.054 sec
TEST: test_all_mode_2            - pass -   3.478 sec
TEST: test_all_mode_3            - pass -   4.482 sec
TEST: test_all_grid_stat_1       - pass -   0.627 sec
TEST: test_all_grid_stat_2       - pass -   0.327 sec
TEST: test_all_grid_stat_3       - pass -   0.919 sec
TEST: test_all_grid_stat_4       - pass -   5.618 sec
TEST: test_all_pb2nc             - pass -   9.015 sec
TEST: test_all_plot_point_obs    - pass -   5.280 sec
TEST: test_all_ascii2nc_1        - pass -   0.260 sec
TEST: test_all_ascii2nc_2        - pass -   0.359 sec
TEST: test_all_madis2nc          - pass -   1.077 sec
TEST: test_all_point_stat        - FAIL -   0.288 sec
/d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/met/share/met/../../bin/point_stat \
      /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/met/data/sample_fcst/2007033000/nam.t00z.awip1236.tm00.20070330.grb \
      /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/test_output/met_test_scripts/pb2nc/sample_pb.nc \
      /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/met/scripts/config/PointStatConfig \
      -outdir /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/test_output/met_test_scripts/point_stat \
      -v 2
DEBUG 1: Default Config File: /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/met/share/met/config/PointStatConfig_default
DEBUG 1: User Config File: /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/met/scripts/config/PointStatConfig
ERROR  : 
ERROR  : Dictionary::lookup_int() -> lookup failed for name "output_flag.orank"
ERROR  : 

ERROR: /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764/MET-feature_1764_point_stat/test/perl/unit.pl unit_met_test_scripts.xml failed.

*** UNIT TESTS FAILED ***
```

**Comment by JohnHalleyGotway** [comment]
Finished reviewing the code changes and after a few commits, the code is all aligned and the point-stat/grid-stat docs are all consistent. Just waiting for the regression test to finish up...

**Review by JohnHalleyGotway — CHANGES_REQUESTED** [review]
The regression test reveals that output has changed in only 2 files:
```
cd /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1764
egrep -i "file1|error:" test_regression.log | egrep -B 1 ERROR: | grep file1
file1: MET-feature_1764_point_stat/test_output/hira/point_stat_HIRA_EMPTY_PROB_CAT_THRESH_360000L_20120410_120000V.stat
file1: MET-feature_1764_point_stat/test_output/hira/point_stat_NCMET_NAM_HMTGAGE_HIRA_360000L_20120410_120000V.stat
```
I found an issue in the setting of the OBS_VALID_BEG and OBS_VALID_END columns.
In point_stat_HIRA_EMPTY_PROB_CAT_THRESH_360000L_20120410_120000V.stat, for the RPS and ECNT line types, the values change.
FROM: 20120410_113000 20120410_123000
TO: 20120410_120000 20120410_120000

After writing the HiRA ORANK lines, reset the OBS_VALID_BEG and OBS_VALID_END columns back to what they had been. Please see this [line](https://github.com/dtcenter/MET/blob/6d7b55c25c022f80d847dde0ef661fec22a97cb1/met/src/tools/core/ensemble_stat/ensemble_stat.cc#L1418) from Ensemble-Stat and this [line](https://github.com/dtcenter/MET/blob/6d7b55c25c022f80d847dde0ef661fec22a97cb1/met/src/tools/core/point_stat/point_stat.cc#L1044) from Point-Stat, and do the same thing after calling write_orank_row(...).

**Comment by sethlinden** [comment]
Hi John, thanks for letting me know about having to to reset OBS_VALID_BEG and OBS_VALID_END after calling write_orank_row. I'll make those changes now.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes.

I re-ran unit_hira.xml with the latest changes from @sethlinden and confirmed that the only differences between the two Point-Stat output files are the addition of the new ORANK lines.

Note that this PR changes the output of the unit tests, so the regression test will fail. @JohnHalleyGotway will merge in these changes and then restart last night's nightly build.

**Pull request merged**